test/common/buffer/owned_impl_test.cc

```diff
@@ -794,188 +783,155 @@ TEST_F(OwnedImplTest, ReserveCommit) {
 
   {
     Buffer::OwnedImpl buffer;
-    // A zero-byte reservation should fail.
-    static constexpr uint64_t NumIovecs = 16;
-    Buffer::RawSlice iovecs[NumIovecs];
-    uint64_t num_reserved = buffer.reserve(0, iovecs, NumIovecs);
-    EXPECT_EQ(0, num_reserved);
-    clearReservation(iovecs, num_reserved, buffer);
+    // A zero-byte reservation should return an empty reservation.
+    {
+      auto reservation = buffer.reserveApproximately(0);
+      EXPECT_EQ(0, reservation.numSlices());
+    }
     EXPECT_EQ(0, buffer.length());
 
     // Test and commit a small reservation. This should succeed.
-    num_reserved = buffer.reserve(1, iovecs, NumIovecs);
-    EXPECT_EQ(1, num_reserved);
-    // The implementation might provide a bigger reservation than requested.
-    EXPECT_LE(1, iovecs[0].len_);
-    iovecs[0].len_ = 1;
-    commitReservation(iovecs, num_reserved, buffer);
+    {
+      auto reservation = buffer.reserveApproximately(1);
+      EXPECT_EQ(1, reservation.numSlices());
+      EXPECT_EQ(16384, reservation.slices()[0].len_);
+      reservation.commit(1);
+    }
     EXPECT_EQ(1, buffer.length());
 
     // Request a reservation that fits in the remaining space at the end of the last slice.
-    num_reserved = buffer.reserve(1, iovecs, NumIovecs);
-    EXPECT_EQ(1, num_reserved);
-    EXPECT_LE(1, iovecs[0].len_);
-    iovecs[0].len_ = 1;
-    const void* slice1 = iovecs[0].mem_;
-    clearReservation(iovecs, num_reserved, buffer);
+    const void* slice1;
+    {
+      auto reservation = buffer.reserveApproximately(1);
+      EXPECT_EQ(1, reservation.numSlices());
+      EXPECT_EQ(1, reservation.slices()[0].len_);
+      slice1 = reservation.slices()[0].mem_;
+    }
 
     // Request a reservation that is too large to fit in the remaining space at the end of
     // the last slice, and allow the buffer to use only one slice. This should result in the
     // creation of a new slice within the buffer.
-    num_reserved = buffer.reserve(4096, iovecs, 1);
-    EXPECT_EQ(1, num_reserved);
-    EXPECT_NE(slice1, iovecs[0].mem_);
-    clearReservation(iovecs, num_reserved, buffer);
+    {
+      auto reservation = buffer.reserveSingleSlice(16384);
+      EXPECT_EQ(16384, reservation.slice().len_);
+      EXPECT_NE(slice1, reservation.slice().mem_);
+    }
 
     // Request the same size reservation, but allow the buffer to use multiple slices. This
     // should result in the buffer creating a second slice and splitting the reservation between the
     // last two slices.
-    num_reserved = buffer.reserve(4096, iovecs, NumIovecs);
-    EXPECT_EQ(2, num_reserved);
-    EXPECT_EQ(slice1, iovecs[0].mem_);
-    clearReservation(iovecs, num_reserved, buffer);
+    {
+      auto reservation = buffer.reserveApproximately(16384);
+      EXPECT_EQ(2, reservation.numSlices());
+      EXPECT_EQ(slice1, reservation.slices()[0].mem_);
```

Comment:
Please add:
EXPECT_EQ(32767, reservation.length());
EXPECT_EQ(16383, reservation.slices()[0].len_);
EXPECT_EQ(16384, reservation.slices()[1].len_);

This is one of the odd cases where reserving 16kb ends up reserving significantly more.